packages/instantsearch.js/src/lib/utils/hydrateSearchClient.ts

```diff
@@ -0,0 +1,132 @@
+// @ts-nocheck (types to be fixed during actual implementation)
+import type { InitialResults, SearchClient } from '../../types';
+
+export function hydrateSearchClient(
+  client: SearchClient,
+  results?: InitialResults
+) {
+  if (!results) {
+    return;
+  }
+
+  // Disable cache hydration on:
+  // - Algoliasearch API Client < v4 with cache disabled
+  // - Third party clients (detected by the `addAlgoliaAgent` function missing)
+
+  if (
+    (!client.transporter || client._cacheHydrated) &&
+    (!client._useCache || typeof client.addAlgoliaAgent !== 'function')
+  ) {
+    return;
+  }
+
+  // Algoliasearch API Client >= v4
+  // To hydrate the client we need to populate the cache with the data from
+  // the server (done in `hydrateSearchClientWithMultiIndexRequest` or
+  // `hydrateSearchClientWithSingleIndexRequest`). But since there is no way
+  // for us to compute the key the same way as `algoliasearch-client` we need
+  // to populate it on a custom key and override the `search` method to
+  // search on it first.
+  if (client.transporter && !client._cacheHydrated) {
+    client._cacheHydrated = true;
+
+    const baseMethod = client.search;
+    client.search = (requests, ...methodArgs) => {
+      const requestsWithSerializedParams = requests.map((request) => ({
+        ...request,
+        params: serializeQueryParameters(request.params),
+      }));
+
+      return client.transporter.responsesCache.get(
+        {
+          method: 'search',
+          args: [requestsWithSerializedParams, ...methodArgs],
+        },
+        () => {
+          return baseMethod(requests, ...methodArgs);
+        }
+      );
+    };
+
+    // Populate the cache with the data from the server
+    client.transporter.responsesCache.set(
+      {
+        method: 'search',
+        args: [
+          Object.values(results).reduce(
+            (acc, result) =>
+              acc.concat(
+                result.results.map((request) => ({
+                  indexName: request.index,
+                  params: request.params,
+                }))
+              ),
+            []
+          ),
+        ],
+      },
+      {
+        results: Object.values(results).reduce(
```

Comment:
Object.values isn't in the polyfill allowlist, either allow it or change the function used (using Object.keys for example)